stm32: Ringbuffer rewrite

liarokapisv opened the pull request:
The current ring buffer implementation is reportedly a bit flaky and hard to audit.

This PR aims to rewrite the ring buffer module with a simpler implementation that is easier to audit and has clear semantics.
The new module includes [stateful property tests](https://proptest-rs.github.io/proptest/proptest/state-machine.html) which can encode and test the semantics of the module to ensure correctness.

I have been testing this implementation with a real-life DSP project, for both the `USARTX` code and the `I2S` codec processing. I of course welcome others to test the PR to get feedback.

One point worth discussing is the semantics of the `new` and `clear` methods for the `WritableDmaRingBuffer`.
After these calls, the ring buffer needs to assume some existing length, otherwise it will `Underrun` on the next `write` call.
Currently, these calls directly force a `length` equal to the `capacity` giving the maximum leeway for the DMA to catchup, but we may want to make these configurable. This has the disadvantage of directly using the previous contents of the underlying buffer at a non user-controllable offset. This is not a great idea due to the lack of user control, but is not fixable without changing the interface.

I am also a bit skeptical about the recently introduced `write_immediate` buffer method.

I think another approach is worth experimenting with, that of directly constructing/resetting an empty `WritableDmaRingBuffer` and having to hydrate with a version of `write` that does not produce an `Underrun` error. 

A possible issue with the PR is that it removes the `get_complete_count` from the `DmaCtrl` trait since it uses the `reset_complete_count` instead. It may be worth reverting this to squeeze some extra performance.

As always, thanks for the great project and I hope this PR is useful.

liarokapisv commented:
@xoviat @rmja Your feedback would be invaluable for this PR.

liarokapisv commented:
Another thing is that the state machine basically stays on Overrun state until the user calls clear. This may be unwanted.

liarokapisv commented:
bender run

Dirbaio commented:
bender run

liarokapisv commented on embassy-stm32/src/dma/ringbuffer/mod.rs:
```diff
@@ -0,0 +1,293 @@
+            .write_index
+            .diff(self.cap(), &mut self.read_index)
+            .try_into()
+            .unwrap();
```
The implicit invariant is that the read index is always behind the write index. It should not happen since we only update the read index up to the diff between it and the write index.

dvdsk commented:
Could the ringbuffer be extended with an `is_empty()` or `len()` member? 

That would make it easy to implement read_ready() for the RingbufferUart and other ringbuffer backed IO

peterkrull commented on embassy-stm32/src/dma/ringbuffer/mod.rs:
```diff
@@ -0,0 +1,293 @@
+        .await
+    }
+
+    fn update_dma_index(&mut self, dma: &mut impl DmaCtrl) {
```
We could also inline the function body wherever it is used, removing this function altogether, same for the WritableDmaRingBuffer.

liarokapisv commented:
I think it makes sense, will do.

peterkrull commented on embassy-stm32/src/dma/ringbuffer/mod.rs:
```diff
@@ -0,0 +1,293 @@
+    pos: usize,
+}
+
+fn pos(cap: usize, dma: &impl DmaCtrl) -> usize {
```
Since this is only used in dma_sync I would just inline it there

liarokapisv commented:
@peterkrull @dvdsk

I simplified the implementation by inlining some methods.
I also renamed `margin` to `len` and split the auto-normalization functionality of `diff` along with the `update-dma-index` to a separate `sync-index` method. This makes the `len` method take a `&self` instead of a `&mut self`.
I also added `auto-clear` functionality: this fast-forwards the index for the read version and gives a buffer worth of leeway for the writer implementation. The latter may need changes, I always welcome any feedback.

dvdsk commented:
> I also renamed `margin` to `len`

Thanks I'll get onto adding ReadReady once this is merged.

peterkrull commented:
> That would make it easy to implement read_ready() for the RingbufferUart and other ringbuffer backed IO

What is the expected behavior of read_ready()? With `len` now being read-only it does not actually fetch whether the hardware DMA state is actually updated. It only returns the state from the last time the writer index was synced.

liarokapisv commented:
> > That would make it easy to implement read_ready() for the RingbufferUart and other ringbuffer backed IO
> 
> What is the expected behavior of read_ready()? With `len` now being read-only it does not actually fetch whether the hardware DMA state is actually updated. It only returns the state from the last time the writer index was synced.

Yea I was thinking this may be problematic, I can revert to a mutating version. The read-onyl version will not be updated until a new read so that doesn't make a lot of sense in retrospect.

liarokapisv commented:
Two main issues that I believe still remain to be properly discussed:

# Overrun recovery

Currently, the read ring-buffer version directly syncs to the current DMA location. I think this is pretty straightforward and covers most use cases properly. 

The write version on the other hand directly syncs `buf.len` steps ahead of the current DMA location. This is kind of arbitrary, it ensures that the user has the most leeway to properly get back on track, but it also adds latency to the subsequent writes. This may also lead to sync issues between parallel read and write calls. 
This is similar to what happens at ring buffer creation, since at that point the write-ring-buffer is full.

# Broken Invariants Detection

The implementation assumes that the RingBuffers are only used while the DMA is both 1) enabled and 2) not modified other than through the `DmaCtrl` interface. If the DMA stream is reset while the buffer is in operation, or if the ringbuffer is used while the stream is not enabled, then these assumptions are broken. 

Some of this can be detected through negative `diff` method results.  At that point, we could either panic (this is what happens currently) or instead propagate a new error type and make this recoverable. This would prevent issues like those caused in #3356 but would also require modifying other ringbuffer clients without any obvious way to recover. Another possible approach is to sneak this error case as part of the `Overrun` error path perhaps also logging the detected broken invariant.

peterkrull commented:
For the invariant thing, I am still most in favor of just returning an error. An `Overrun` error would be fine for me, but perhaps someone more experienced can chime in on whether a new type of error is warranted. Then add a doc comment to `ReadableDmaRingBuffer::clear()` that it should only be called after the DMA is already enabled/started for the most predictable behavior.

Another little thing, dma_sync can be simplified by assigning directly to the `complete_count` and `pos`. I would also rename completion_count to complete_count since that is what it is called elsewhere.
```rust
fn dma_sync(&mut self, cap: usize, dma: &mut impl DmaCtrl) {
    let first_pos = cap - dma.get_remaining_transfers();
    self.complete_count += dma.reset_complete_count();
    self.pos = cap - dma.get_remaining_transfers();

    // If the latter call to get_remaining_transfers() returned a smaller value than the first, the dma
    // has wrapped around between calls and we must check if the complete count also incremented.
    if self.pos < first_pos {
        self.complete_count += dma.reset_complete_count();
    }
}
```
 
As for the WritableDmaRingBuffer I have only started looking at it now

peterkrull commented:
Apart from the initial added delay, what is the drawback of just giving the most leeway when starting/clearing the WritableDmaRingBuffer? Presumably whoever decides the length of the DMA buffer takes into account the acceptable delay of the whole system, and the worst-case execution time. If a possible latency of `dt*dma_buf.len()` is too long, then shouldn't the buffer just be shorter? Also, if the user did in fact overrun just prior, then they might need all the help they can get to catch up.

Unless there is some clear immediate need for a custom initial offset, defaulting to the maximum is probably fine.

liarokapisv commented:
My main problematic case goes something like this: 
User in a loop repeatedly calls `read_exact.await`, processes, then `write_exact.await`. Clearing will then also affect the read call potentially leading to cascading overrun issues. 

This is recoverable, just requires some thought in the application logic. I agree that giving the max length is a good default hence why I made it so,  just thought I should get some feedback first before committing.

liarokapisv commented:
Also some bikesheding, should clear be sync or something? Clear makes complete sense for the read version whefe the len will be reset to 0 but is perhaps counterintuitive for the write version which skips buf.len steps ahead of the dma.

peterkrull commented:
Could both not just be called reset? The doc comment already describes them as a reset

liarokapisv commented:
> Could both not just be called reset? The doc comment already describes them as a reset

I guess reset is ok. It matches the behavior at construction as well.

rmja commented:
Hi. Sorry for not coming with feedback to this before now. I am absolutely fine with this being updated to a more robust implementation - although I have not seen the flaky behavior that has been reported, at least not on stm32f4.

One thing that I cannot immediately see in this implementation that was included before is the protection against buffer overruns in the case where the overrun is large and spans one or more entire dma buffers. Now it seems that we are only incrementing a local `complete_count` based on the result of `dma.reset_complete_count()` instead of using the `get_complete_count` to determine if multiple resets of NDTR has happend by the dma controller in between read's?

I might be missing if the above is actually the case - I have been off rust dev. for quite some time now.

peterkrull commented:
> [rmja](https://github.com/rmja): One thing that I cannot immediately see in this implementation that was included before is the protection against buffer overruns in the case where the overrun is large and spans one or more entire dma buffers.

That protection should be included here, since any positive difference larger than the capacity of the buffer is an overrun condition.
https://github.com/embassy-rs/embassy/blob/a8328fd7c20d5d03b2cc438a3ac94c6ba411d632/embassy-stm32/src/dma/ringbuffer/mod.rs#L214

> [rmja](https://github.com/rmja): Now it seems that we are only incrementing a local complete_count based on the result of dma.reset_complete_count() instead of using the get_complete_count to determine if multiple resets of NDTR has happend by the dma controller in between read's?

The underlying atomic complete_count counts up like before in the interrupt handler. Now we just "take" (and reset) the atomic value to keep it as a local state within the circular buffer. So the complete_count atomic just represents the number of times the buffer filled since the last time `reset_complete_count` was called.

---

> [liarokapisv](https://github.com/liarokapisv): Furthermore, I also applied the Usart panic fix to the Adc as well

You should probably also change the adc teardown to only pause the DMA and not fully stop it.

https://github.com/embassy-rs/embassy/blob/a8328fd7c20d5d03b2cc438a3ac94c6ba411d632/embassy-stm32/src/adc/ringbuffered_v2.rs#L247

liarokapisv commented:
> You should probably also change the adc teardown to only pause the DMA and not fully stop it.
> 
> https://github.com/embassy-rs/embassy/blob/a8328fd7c20d5d03b2cc438a3ac94c6ba411d632/embassy-stm32/src/adc/ringbuffered_v2.rs#L247

Noted and fixed. I am not sure if I need to do any other modifications.

rmja commented:
@peterkrull thanks, I see that now. The implementation looks good from here! Thank you all for putting effort into making this more robust.

liarokapisv commented:
It appears that there exists a fundamental issue with the `complete_count` variable. It is possible, especially at high speeds that we never read a `complete_count` increase quickly after a wraparound. This is because something may block the dma complete interrupt from firing before any thread-level code reads the position.

I have a working patch ready that addresses this issue by performing a blocking wait if the positions are inconsistent:

```rust
    fn dma_sync(&mut self, cap: usize, dma: &mut impl DmaCtrl) {
        let original_pos = self.pos;
        let first_pos = cap - dma.get_remaining_transfers();
        let first_count = dma.reset_complete_count();
        self.complete_count += first_count;
        self.pos = cap - dma.get_remaining_transfers();

        let second_count = if (first_pos < original_pos || self.pos < first_pos) && first_count == 0 {
            loop {
                let count = dma.reset_complete_count();
                if count != 0 {
                    break count;
                }
            }
        } else {
            0
        };

        self.complete_count += second_count;
    }
```

It works in my case in that it fixes all overrun issues but it makes the call unbounded and is especially dangerous if called from IRQ context.
I don't expect the blocking call to be large in practice.

I have a feeling that this should be properly handled from within the `dma_bdma` module by tracking the transfer complete flag
and properly fixing the reported completion counts.

Any suggestions on fixing this is welcome.

liarokapisv commented:
My best solution so far is to pretend to still be on the previous cycle if we can't detect a count update.

```rust
fn dma_sync(&mut self, cap: usize, dma: &mut impl DmaCtrl) {
    let new_pos = cap - dma.get_remaining_transfers();
    let count_diff = dma.reset_complete_count();
    self.pos = if new_pos < self.pos && count_diff == 0 {
        cap - 1     
    } else {
        new_pos
    };
    self.complete_count += count_diff;
}
```

The main idea is that the dma position is always an approximation, there are always more samples to come, and we will get re-polled anyway, so we just pretend to still be on the last cycle. 

My only worry about it is that it /may/ add some latency to `read_exact` and force it to way until the next waker update. 
I don't think this is a really an issue because it would probably only happen at the end of the current `read_exact` cycle where it would be starved of new samples in any case.

liarokapisv commented:
Turns out the order of the first two lines is important.
The correct version is:

```rust
  fn dma_sync(&mut self, cap: usize, dma: &mut impl DmaCtrl) {
      let count_diff = dma.reset_complete_count();
      let pos = cap - dma.get_remaining_transfers();
      self.pos = if pos < self.pos && count_diff == 0 {
          cap - 1
      } else {
          pos
      };

      self.complete_count += count_diff;
  }
```

The count needs to be read first to ensure we don't get a +capacity jump.

liarokapisv commented:
I think I have addressed most of my issues. 
I still would like some feedback on the current `ringbuffer Error -> Overrun` conversion in the case of `DmaUnsynced` error. With the last update, it only logs an error message and silently converts to `Overrun`. I was thinking it may be worth panicking since it implies an `embassy` driver issue.

peterkrull commented:
I would very much prefer not to panic. It would really suck to have my drone fall out of the sky just because one of my RC packets (for whatever reason) caused invariant we have tried to model in the buffer, not to be upheld. Especially when one lost packet and a DMA restart later is all it takes to get back on track. It would be another thing if the entire peripheral locked up unrecoverably, but that is not the case so I prefer the error.

Also can the if..if..else at (and same for the writer)

https://github.com/embassy-rs/embassy/pull/3336/commits/4d2be7a905062d343e6e6ca3232bba226ebf54af#diff-79e995a8dc7e2d1878dc6f7a4f665c9aea320f6f0e589e178769ed329a570337R113-R118

be combined into an if..ifelse..else?

liarokapisv commented:
@peterkrull done.

Dirbaio commented:
bender run

Dirbaio commented:
bender run

liarokapisv commented:
@Dirbaio I don't think I have anything else to address. I am open to feedback but otherwise I think this mostly does it. @peterkrull Do you agree?

peterkrull commented:
I do not have anything else to add

liarokapisv commented:
There is an issue with the (unchanged) `read/write_exact` calls: They only poll once before yielding (probably to ensure forward progress). If one writes less than half the dma buffer and there are no spurious wake-ups it is almost guaranteed that the user will get overuns because the method simply does not get to be run enough times to fill the buffer.
